src/renderers/shared/fiber/ReactFiberHostContext.js

```diff
@@ -0,0 +1,151 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactFiberHostContext
+ * @flow
+ */
+
+'use strict';
+
+import type { Fiber } from 'ReactFiber';
+import type { HostConfig } from 'ReactFiberReconciler';
+
+export type HostContext<C, CX> = {
+  getRootHostContainer() : C,
+  getHostContext() : CX | null,
+
+  pushHostContext(fiber : Fiber) : void,
+  popHostContext(fiber : Fiber) : void,
+
+  pushHostContainer(container : C) : void,
+  popHostContainer() : void,
+  resetHostContainer() : void,
+};
+
+module.exports = function<T, P, I, TI, C, CX>(
+  config : HostConfig<T, P, I, TI, C, CX>
+) : HostContext<C, CX> {
+  const {
+    getChildHostContext,
+  } = config;
+
+  type ContainerState = {
+    rootInstance : C,
+    contextFibers : Array<Fiber | null> | null,
+    contextValues : Array<CX | null> | null,
+    contextDepth: number
+  };
+
+  function createContainerState(rootInstance) {
+    return {
+      rootInstance,
+      contextFibers: null,
+      contextValues: null,
+      contextDepth: -1,
+    };
+  }
+
+  // If we meet any portals, we'll go deeper.
+  let containerStack : Array<ContainerState> = [];
```

Comment:
If you have a separate variable for the top value, it becomes faster to read since we do a lot more reads than we do push/pop. Same for the context.